Improvements related to git, prompts, usability, refactoring, tests

olehermanse opened the pull request:
- Suppressed noisy hint output when running git init for cfbs init
- Switched to using CFEngine Build project terminology
- Made the git commit output more easily distinguishable
- Made the Initialized message a bit more informative
- Moved prompting code to separate file
- Added a prompt for using default commit message by pressing enter
- Moved git related code to git.py
- Made print_module_info() "private"
- Made download_dependencies() "private"
- Moved longest_module_name to CFBSConfig()
- Moved _item_index() to utils
- Renamed definition to config for consistency
- Fixed inconsistency with index vs index_path in cfbs.json
- Moved build related code to build.py
- Renamed functions related to building the policy set
- Replaced distutils.spawn.find_executable() with shutil.which()
- Converted test_add_folder.py to a shell tests
- Fixed issue with running cfbs info outside a project
- Changed capitalization in cfbs info output
- Replaced test_showinfo.py with shell based test
- Reformatted files with black
- Prevented cfbs from trying to make commits when nothing was added

vpodzime commented on cfbs/git.py:
```diff
@@ -1,7 +1,9 @@
+from functools import partial
 from subprocess import check_call, check_output, run, PIPE, DEVNULL, CalledProcessError
 from cfbs.prompts import YES_NO_CHOICES, prompt_user
+from cfbs.cfbs_config import CFBSConfig
```
I intentionally tried to keep the `CFBSConfig`-dependent (and thus cfbs-specific) code outside of this module to make it reusable. Especially the `commit_after_command` is very specific to the commands handling.

olehermanse replied:
It doesn't belong in commands and to me it makes the most sense to have all the git code in one place, but I can split it in two, `git.py` and `git_magic.py`.

olehermanse replied:
Added a new commits which makes `git.py` reusable and puts the rest of the code in `git_magic.py`.

vpodzime commented on cfbs/commands.py:
```diff
@@ -506,7 +506,7 @@ def validate_command():
 
 
-def init_build_folder(config):
+def init_build_folder():
```
Looks like a piece of change from the next commit, but most likely not worth moving to it.

olehermanse replied:
Thanks, I fixed it :)

vpodzime reviewed (commented):
There are other commands that should use the `CFBSReturnWithoutCommit` exception mechanism (I did `return 2` in those cases).

olehermanse commented on cfbs/git.py:
```diff
@@ -75,17 +75,27 @@ def git_init(user_name=None, user_email=None, description=None):
+            "The default commit message is '{}' - edit it?".format(commit_msg),
+            choices=YES_NO_CHOICES,
+            default="no",
+        )
```
We can work on making the prompts nicer. Always pressing enter for defaults is a good pattern. I know it's easier for you, but I'm trying to make it as easy as possible for new users.

olehermanse commented on cfbs/git.py:
```diff
@@ -75,17 +75,27 @@ def git_init(user_name=None, user_email=None, description=None):
+            choices=YES_NO_CHOICES,
+            default="no",
+        )
+        edit_commit_msg = ans in ("yes", "y")
```
Thanks, this was copy paste from another place, so I've added another commit to fix there as well.

olehermanse commented on cfbs/git.py:
```diff
@@ -57,7 +57,8 @@ def git_init(user_name=None, user_email=None, description=None):
     try:
-        check_call(["git", "init"])
+        # Suppress noisy hint output on stderr:
+        check_call(["git", "init"], stderr=DEVNULL)
```
To me, this is a big improvement in terms of reading the output. They would get the same hint if they run `git init` somewhere, and if I'm running `cfbs init` I don't really care that there is an option to set another default branch name.

olehermanse commented:
> There are other commands that should use the `CFBSReturnWithoutCommit` exception mechanism (I did `return 2` in those cases).

@vpodzime addressed comments and added the new exception in other places as well.

olehermanse commented on cfbs/commands.py:
```diff
@@ -289,7 +289,7 @@ def init_command(index_path=None, non_interactive=False) -> int:
             return 1
 
-    print("Initialized")
+    print("Initialized an empty project called '{}' in {}".format(name, cfbs_filename()))
```
It can be whatever you want it to be 😅 The reason was probably that the name is user input, can include spaces, which the filename is just `cfbs.json`. I can add the quotes 🚀

olehermanse replied:
Done.

olehermanse commented on cfbs/cfbs_config.py:
```diff
@@ -273,14 +274,23 @@ def add_command(
+        if len(self["build"]) == before_adding:
+            # Not an error, we just want to exit successfully without
+            # making a git commit
+            raise CFBSReturnWithoutCommit(0)
```
What you are suggesting seems hacky to me.

These are commands which have a return code for success or failure, which goes all the way up to be the exit code of the program. All of the commands are the same (even the ones which don't commit). The git commit decorator wraps around some of the commands, but shouldn't significantly change their signature (parameters and return value). For commands which make changes, the normal scenario is that when successful, a commit should be made. In the exceptional scenario where the command succeeded but didn't make any changes, it is handled with an exception. That seems appropriate to me.

Another advantage of using an exception is that you "jump" directly from where you know you're done, to the catch, even if you are inside another function, you don't have to shuffle return values all the way up.

olehermanse replied:
@larsewi do you mean using exceptions for this in general or this specific place?

larsewi replied:
No, it's fine ;)

olehermanse commented on cfbs/git.py:
```diff
@@ -86,6 +86,8 @@ def git_commit(commit_msg, edit_commit_msg=True, scope="all"):
 
     """
 
+    print("Committing using git:\n")
```
Yes, that's a good idea, but since we are not using colors yet, I'll leave it for another PR.

larsewi reviewed (approved):
LGTM, thanks for all the refactoring you've been doing lately

vpodzime reviewed (approved):
Code looks good to me.
